tests/test_cases/test_axi4/test_axi4.py

```diff
@@ -0,0 +1,451 @@
+                "was expecting {:#x} ({:#x} with {:#x} as strobe and {:#x} as "
+                "previous value)"
+                .format(read_values[i].integer, i + 1, burst_length, address,
+                        expected_values[i], write_values[i], strobes[i],
+                        previous_values[i].integer))
+
+
+@cocotb.test()
+async def test_fixed_wrap_burst(dut, burst_length=16):
+    """Test FIXED and WRAP read/writes"""
+
+    axim = AXI4Master(dut, AXI_PREFIX, dut.clk)
+    _, data_width, ram_start, ram_stop = get_parameters(dut)
+
+    if data_width == 1:
+        raise TestFailure("This test cannot be executed with data_width == 1")
+
+    await setup_dut(dut)
+
+    base = randrange(ram_start, ram_stop, 4096)
+    offset = randrange(0, 4096 - (burst_length - 1) * data_width, data_width)
+    address = base + offset
+
+    for burst in (AXIBurst.FIXED, AXIBurst.WRAP):
+
+        write_values = \
+            [randrange(0, 2**(data_width * 8)) for i in range(burst_length)]
+
+        await axim.write(address, write_values, burst=burst)
+
+        if burst is AXIBurst.FIXED:
+            # A FIXED write on a memory is like writing the last element,
+            # reading it with a FIXED burst returns always the same value.
+            expected_values = [write_values[-1]] * burst_length
+        else:
+            # Regardless of the boundary, reading with a WRAP from the same
+            # address with the same length returns the same sequence.
+            # This RAM implementation does not support WRAP bursts and treats
+            # them as INCR.
+            expected_values = write_values
+
+        read_values = await axim.read(address, burst_length, burst=burst)
+
+        for i in range(len(read_values)):
+            if expected_values[i] != read_values[i]:
+                raise TestFailure(
+                    "Read {:#x} at beat {}/{} of {} burst with starting "
+                    "address {:#x}, but was expecting {:#x})"
+                    .format(read_values[i].integer, i + 1, burst_length,
+                            burst.name, address, expected_values[i]))
+
+
+@cocotb.test()
+async def test_size(dut):
+    """Test reads/writes when size = data_width // 2"""
+
+    axim = AXI4Master(dut, AXI_PREFIX, dut.clk)
+    _, data_width, ram_start, ram_stop = get_parameters(dut)
+
+    if data_width == 1:
+        raise TestFailure("This test cannot be executed with data_width == 1")
+
+    await setup_dut(dut)
+
+    burst_length = (randrange(2, 256) // 2) * 2
+    address = ram_start
+    write_values = \
+        [randrange(0, 2**(data_width * 8 // 2)) for i in range(burst_length)]
+
+    await axim.write(address, write_values, size=data_width // 2)
+
+    # This slave does not support unaligned transfers or size different than
+    # the data bus width and simply drops the address' least significant bits.
+    # So, for each pair of writes, the second one overwrites the first one.
```

Comment:
AXI does not support unaligned transfers.  The lower-order bits are only used for narrow bursts.

Reply:
You are right, bad wording. And I did not handle correctly narrow bursts anyways, so this has to be fixed. Thanks.